src/utils/board/board.ts

```diff
@@ -0,0 +1,207 @@
+    return currentPiece.piece?.color === selectedPiece.piece?.color;
+};
+
+const isSamePiece = (currentPiece: ICell, selectedPiece: ICell): boolean => {
+    return currentPiece.row === selectedPiece.row && currentPiece.col === selectedPiece.col;
+};
+
+const getAllowedDirections = (boardSize: number): number[] => {
+    const firstAllowedDirection = boardSize - 1;
+    const secondAllowedDirection = boardSize + 1;
+
+    return [firstAllowedDirection, secondAllowedDirection];
+};
+
+const isShouldBecomeKing = (currentPiece: ICell, selectedPiece: ICell): boolean => {
+    if (!currentPiece) {
+        return false;
+    }
+    if (selectedPiece.row === 0 && currentPiece.piece?.color === PieceColor.White) {
+        return true;
+    }
+
+    return selectedPiece.row === BOARD_SIZE - 1 && currentPiece.piece?.color === PieceColor.Black;
+};
+
+const movePiece = (
+    board: ICell[],
+    currentPiece: ICell,
+    selectedPiece: ICell,
+    updateBoard: boolean,
+    activeSide: PieceColor,
+    setBoard: (board: ICell[]) => void,
+    setCurrentPiece: (piece: ICell | null) => void,
+    setUpdateBoard: (updateBoard: boolean) => void,
+    setActiveSide: (activeSide: PieceColor) => void,
+): void => {
+    const newBoard = Object.assign(board);
+    const oldPosition = newBoard.findIndex((item: ICell) => (
+        item.row === currentPiece?.row && item.col === currentPiece?.col)
+    );
+    const newPosition = newBoard.findIndex((item: ICell) => (
+        item.row === selectedPiece?.row && item.col === selectedPiece?.col)
+    );
+
+    newBoard[newPosition] = {
+        ...newBoard[newPosition],
+        piece: newBoard[oldPosition].piece,
+    };
+
+    newBoard[oldPosition] = {
+        ...newBoard[oldPosition],
+        piece: null,
+    };
+
+    if (
+        currentPiece.piece?.state !== PieceState.King &&
+        isShouldBecomeKing(currentPiece, selectedPiece)
+    ) {
+        newBoard[newPosition] = {
+            ...newBoard[newPosition],
+            piece: {
+                ...newBoard[newPosition].piece,
+                state: PieceState.King
+            },
+        };
+    }
+
+    setBoard(newBoard);
+    setCurrentPiece(null);
+    setUpdateBoard(!updateBoard);
+    switchSide(activeSide, setActiveSide);
+};
+
+const beatPiece = (
+    board: ICell[],
+    currentPiece: ICell,
+    selectedPiece: ICell,
+    updateBoard: boolean,
+    activeSide: PieceColor,
+    beatByWhite: ICell[],
+    beatByBlack: ICell[],
+    setCurrentPiece: (piece: ICell | null) => void,
+    setActiveSide: (activeSide: PieceColor) => void,
+    setBoard: (board: ICell[]) => void,
+    setUpdateBoard: (updateBoard: boolean) => void,
+    setCanBeatAgain: (canBeatAgain: boolean) => void,
+    canBeatPiece?: boolean,
+): void => {
+    const newBoard = Object.assign(board);
+
+    const currentPosition = newBoard.findIndex((item: ICell) => (
+        item.row === currentPiece?.row && item.col === currentPiece?.col)
```

Comment:
Can currentPiece not be defined?

Reply:
No, it can't. I wrote that just for safety. Or in this case it's excess? I mean can someone be confusing by this syntax? If we will get currentPiece for sure than why do we need to check for it?

Reply:
Typescript has to be implicit. If your interface states that this is defined, you shouldn't re-check it by adding `?`.
If you are not sure if some of the fields of the interface can be defined or not, point that in the interface type.

Otherwise it's confusing